File: 08_Tries/03_Word_Search_II/0212-word-search-ii.cpp
/*
Problem: LeetCode 212 - Word Search II

Description:
Given an m x n board of characters and a list of words, find all words in the board.
Each word must be constructed from letters of sequentially adjacent cells, where adjacent cells are horizontally or vertically neighboring.
The same letter cell may not be used more than once in a word.

Intuition:
To find all the words in the board, we can use the Trie data structure to efficiently search for each word. We perform a depth-first search (DFS) starting from each cell on the board, checking if the current sequence of characters forms a valid word in the Trie.

Approach:
1. TrieNode:
   - Define a TrieNode class that represents each node in the Trie.
   - Each TrieNode has an unordered_map to store the child nodes, representing the lowercase alphabets.
   - Each TrieNode also has a boolean flag to indicate if it represents a complete word.

2. Build the Trie:
   - Construct a Trie by inserting each word from the given list into the Trie.

3. DFS Search:
   - Perform a depth-first search (DFS) starting from each cell on the board.
   - At each cell, check if the current character exists in the Trie and move to the corresponding child node.
   - Mark the current cell as visited.
   - If the current node represents a complete word, add it to the result.
   - Recursively explore the neighboring cells (up, down, left, right).
   - Backtrack by unmarking the current cell and removing the last character from the current sequence.

4. Word Search II:
   - Initialize an empty vector to store the found words.
   - Iterate through each cell on the board and perform the DFS search.
   - Return the found words as the result.

Time Complexity:
- Building the Trie: O(m), where m is the total number of characters in all words.
- DFS Search: O((m*n)*3^l), where m and n are the dimensions of the board and l is the average length of the words.

Space Complexity:
- The space complexity is O(m), where m is the total number of characters in all words (used for constructing the Trie).
- The space complexity of the recursive stack for DFS is O(l), where l is the maximum length of the words.
*/

class TrieNode {
  public:
    bool isWord;
    unordered_map<char, TrieNode *> children; // Map to store the child nodes

    TrieNode() {
        isWord = false;
    }
};

class Solution {
  public:
    vector<string> findWords(vector<vector<char>> &board, vector<string> &words) {
        TrieNode *root = buildTrie(words); // Build the Trie
        int rows = board.size();
        int cols = board[0].size();
        vector<string> result;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                string currentWord = ""; // Initialize the current word for each cell
                dfs(board, i, j, root, currentWord, result); // Perform DFS search
            }
        }

        return result;
    }

  private:
    TrieNode *buildTrie(vector<string> &words) {
        TrieNode *root = new TrieNode();

        for (string &word : words) {
            TrieNode *node = root;

            for (char c : word) {
                if (node->children.find(c) == node->children.end()) {
                    node->children[c] = new TrieNode();
                }

                node = node->children[c];
            }

            node->isWord = true;
        }

        return root;
    }

    void dfs(vector<vector<char>> &board, int row, int col, TrieNode *node, string &currentWord, vector<string> &result) {
        if (row < 0 || row >= board.size() || col < 0 || col >= board[0].size() || board[row][col] == '#') {
            return;
        }

        char c = board[row][col];

        if (node->children.find(c) == node->children.end()) {
            return;
        }

        node = node->children[c];
        currentWord += c;

        if (node->isWord) {
            result.push_back(currentWord); // Add the found word to the result
            node->isWord = false; // Mark the word as visited
        }

        board[row][col] = '#'; // Mark the current cell as visited
        // Explore the neighboring cells (up, down, left, right)
        dfs(board, row - 1, col, node, currentWord, result);
        dfs(board, row + 1, col, node, currentWord, result);
        dfs(board, row, col - 1, node, currentWord, result);
        dfs(board, row, col + 1, node, currentWord, result);
        board[row][col] = c; // Backtrack: unmark the current cell
        currentWord.pop_back(); // Remove the current character from the current word
    }
};

File: 08_Tries/03_Word_Search_II/0212-word-search-ii.rs
/*
Problem: LeetCode 212 - Word Search II

Key Idea:
The key idea is to use a Trie data structure to efficiently store the given words and then perform a depth-first search (DFS) on the board to find words that match the Trie structure.

Approach:
1. Create a Trie structure with methods for inserting words and tracking word solutions. Each Trie node has 26 children (one for each lowercase letter) and a flag to mark the end of a word.
2. Insert all the words from the given list into the Trie.
3. Initialize an empty result vector `ans` to store the found words.
4. Iterate through each cell of the board and call the DFS function:
   - In the DFS function, mark the current cell as visited by replacing its character with a space.
   - Check if there's a Trie node corresponding to the character at the current cell.
   - If a word solution is found in the Trie, add it to the result vector `ans`.
   - Recursively explore the neighboring cells (up, down, left, right) that have valid characters and Trie nodes associated with them.
   - Restore the original character at the current cell.
5. After processing all cells on the board, return the `ans` vector containing the found words.

Time Complexity:
- Building the Trie from the given words takes O(W*L), where W is the total number of characters in all words, and L is the average word length.
- The DFS operation on the board has a worst-case time complexity of O(N*M*4^L), where N and M are the dimensions of the board, and L is the maximum word length. This is because, in the worst case, the algorithm explores four neighboring cells at each step for each cell on the board.

Space Complexity:
- The space complexity of the Trie structure is O(W*L), where W is the total number of characters in all words, and L is the average word length.
- The space complexity of the DFS stack is O(L), where L is the maximum word length.
*/

struct Trie {
    children: Vec<Option<Trie>>,
    is_solution: Option<String>,
}

impl Trie {
    fn new() -> Self {
        Trie {
            children: (0..26).map(|_| None).collect(),
            is_solution: None,
        }
    }

    fn insert(&mut self, word: String) {
        let mut curr = self;
        for byte in word.bytes() {
            let index = (byte - b'a') as usize;
            curr = curr.children[index].get_or_insert_with(Trie::new);
        }
        curr.is_solution = Some(word);
    }
}

impl Solution {
    pub fn find_words(mut board: Vec<Vec<char>>, words: Vec<String>) -> Vec<String> {
        let mut trie = Trie::new();
        for word in words {
            trie.insert(word);
        }
        let (n, m) = (board.len(), board[0].len());
        let mut ans = Vec::new();
        for i in 0..n {
            for j in 0..m {
                Self::dfs(&mut board, &mut trie, &mut ans, i, j);
            }
        }
        ans
    }

    fn dfs(board: &mut Vec<Vec<char>>, trie: &mut Trie, ans: &mut Vec<String>, i: usize, j: usize) {
        let c = board[i][j];
        board[i][j] = ' ';
        if let Some(next) = trie.children[(c as u8 - b'a') as usize].as_mut() {
            if let Some(word) = next.is_solution.take() {
                ans.push(word);
            }
            for &(x, y) in &[
                (i + 1, j),
                (i.wrapping_sub(1), j),
                (i, j + 1),
                (i, j.wrapping_sub(1)),
            ] {
                if x < board.len()
                    && y < board[0].len()
                    && board[x][y] != ' '
                    && next.children[(board[x][y] as u8 - b'a') as usize].is_some()
                {
                    Self::dfs(board, next, ans, x, y);
                }
            }
        }
        board[i][j] = c;
    }
}

/*
// Another Solution

use std::collections::HashMap;

type TrieNode = HashMap<char, Trie>;

#[derive(Default)]
struct Trie {
    children: TrieNode,
    word: Option<String>,
}

impl Trie {
    fn new() -> Self {
        Trie {
            children: TrieNode::new(),
            word: None,
        }
    }
}

impl Solution {
    pub fn find_words(mut board: Vec<Vec<char>>, words: Vec<String>) -> Vec<String> {
        let letter_freq = board.iter().flatten().fold(HashMap::new(), |mut acc, &ch| {
            *acc.entry(ch).or_insert(0) += 1;
            acc
        });

        let mut root = Trie::default();
        for word in words {
            let should_reverse = match (word.chars().next(), word.chars().last()) {
                (Some(first), Some(last)) => {
                    letter_freq.get(&first).unwrap_or(&0) > letter_freq.get(&last).unwrap_or(&0)
                }
                _ => false,
            };

            let processed_word: String = if should_reverse {
                word.chars().rev().collect()
            } else {
                word.clone()
            };

            let mut node = &mut root;
            for letter in processed_word.chars() {
                node = node.children.entry(letter).or_insert_with(Trie::default);
            }
            node.word = Some(word);
        }

        let mut matched_words = Vec::new();
        let (row_num, col_num) = (board.len(), board[0].len());

        fn backtracking(
            row: usize,
            col: usize,
            parent: &mut Trie,
            board: &mut Vec<Vec<char>>,
            matched_words: &mut Vec<String>,
            row_num: usize,
            col_num: usize,
        ) {
            let letter = board[row][col];
            let curr_node = match parent.children.get_mut(&letter) {
                Some(node) => node,
                None => return,
            };

            if let Some(word_match) = curr_node.word.take() {
                matched_words.push(word_match);
            }

            board[row][col] = '#';

            const ROW_ADJUST: [isize; 4] = [-1, 0, 1, 0];
            const COL_ADJUST: [isize; 4] = [0, 1, 0, -1];

            for i in 0..4 {
                let new_row = row as isize + ROW_ADJUST[i];
                let new_col = col as isize + COL_ADJUST[i];

                if new_row >= 0
                    && new_row < row_num as isize
                    && new_col >= 0
                    && new_col < col_num as isize
                {
                    backtracking(
                        new_row as usize,
                        new_col as usize,
                        curr_node,
                        board,
                        matched_words,
                        row_num,
                        col_num,
                    );
                }
            }

            board[row][col] = letter;

            if curr_node.children.is_empty() && curr_node.word.is_none() {
                parent.children.remove(&letter);
            }
        }

        for (row, col) in (0..row_num).flat_map(|r| (0..col_num).map(move |c| (r, c))) {
            if root.children.contains_key(&board[row][col]) {
                backtracking(
                    row,
                    col,
                    &mut root,
                    &mut board,
                    &mut matched_words,
                    row_num,
                    col_num,
                );
            }
        }

        matched_words
    }
}
*/
